app/controllers/tours_controller.rb

```diff
@@ -28,17 +28,262 @@ def show
+            name
+          }
+        }
+      }
+    GRAPHQL
+
+    @tour = results.data.tour
   end
 
   def create
+    query = create_params
+    begin
+      results = @smart_village.query query
+    rescue Graphlient::Errors::GraphQLError => e
+      flash[:error] = e.errors.messages["data"].to_s
+      @tour = new_tour
+      render :new
+      return
+    end
+    new_id = results.data.create_tour.id
+    flash[:notice] = "Tour wurde erstellt"
+    redirect_to edit_tour_path(new_id)
   end
 
   def update
+    old_id = params[:id]
+    query = create_params
+    logger.warn(query)
+
+    begin
+      results = @smart_village.query query
+    rescue Graphlient::Errors::GraphQLError => e
+      flash[:error] = e.errors.messages["data"].to_s
+      redirect_to edit_tour_path(old_id)
+      return
+    end
+
+    new_id = results.data.create_tour.id
+
+    if new_id.present? && new_id != old_id
+      # Nach dem Erstellen des neuen Datensatzes wird der alte gelöscht
+
+      destroy_results = @smart_village.query <<~GRAPHQL
+        mutation {
+          destroyRecord(
+            id: #{old_id},
+            recordType: "Tour"
+          ) {
+            id
+            status
+            statusCode
+          }
+        }
+      GRAPHQL
+
+      redirect_to edit_tour_path(new_id)
+    else
+      flash[:error] = "Fehler: #{results.errors.inspect}"
+      redirect_to edit_tour_path(old_id)
+    end
   end
 
   def destroy
+    results = @smart_village.query <<~GRAPHQL
+      mutation {
+        destroyRecord(
+          id: #{params["id"]},
+          recordType: "Tour"
+        ) {
+          id
+          status
+          statusCode
+        }
+      }
+    GRAPHQL
+
+    if results.try(:data).try(:destroy_record).try(:status_code) == 200
+      flash["notice"] = "Eintrag wurde gelöscht"
+    else
+      flash["notice"] = "Fehler: #{results.errors.inspect}"
+    end
+    redirect_to tours_path
   end
+
+  private
+
+    def new_tour
+      OpenStruct.new(
+        addresses: [OpenStruct.new],
+        contact: OpenStruct.new(web_urls: [OpenStruct.new]),
+        web_urls: [OpenStruct.new],
+        operating_company: OpenStruct.new(
+          web_urls: [OpenStruct.new],
+          contact: OpenStruct.new(web_urls: [OpenStruct.new]),
+          address: OpenStruct.new
+        ),
+        media_contents: [OpenStruct.new(source_url: OpenStruct.new)]
+      )
+    end
+
+    def create_params
+      @tour_params = params.require(:tour).permit!
+      convert_params_for_graphql
+      Converter::Base.new.build_mutation("createTour", @tour_params)
+    end
+
+    def convert_params_for_graphql
+      # Convert has_many categories
+      if @tour_params["categories"].present?
+        categories = []
+        @tour_params["categories"].each do |_key, category|
+          next if category.blank?
+
+          categories << category
+        end
+        @tour_params["categories"] = categories
+      end
+
+      # Convert length_km to String
```

Comment:
...to integer meinst du wohl?

Reply:
Ja klar, da hatte ich einen Denkfehler. Ist angepasst